src/components/triggersAndActions.ts

```diff
@@ -78,55 +89,103 @@ export const loadScheduledActions = async function (
 }
 
 export async function processTrigger(payload: TriggerPayload) {
+  const { id: trigger_id, trigger, table, record_id } = payload
+
+  // Deduce template ID -- different for each triggered table
+  const templateId = await DBConnect.getTemplateId(table, record_id)
+
   // Get Actions from matching Template
-  const result = await PostgresDB.getActionPluginsByTemplate(payload.table, {
-    record_id: payload.record_id,
-    trigger: payload.trigger,
-  })
+  const result = await DBConnect.getActionsByTemplateId(templateId, trigger)
+
   // Filter out Actions that don't match the current condition
-  const actions: ActionInTemplate[] = []
+  // and separate into Sequential and Async actions
+  const actionsSequential: ActionSequential[] = []
+  const actionsAsync: ActionInTemplate[] = []
 
   for (const action of result) {
     const condition = await evaluateExpression(action.condition, {
       objects: [payload],
-      pgConnection: PostgresDB,
+      pgConnection: DBConnect,
     })
-    if (condition) actions.push(action)
+    if (condition) {
+      if (action.sequence) actionsSequential.push(action as ActionSequential)
+      else actionsAsync.push(action)
+    }
   }
 
-  // Evaluate parameters for each Action
-  for (const action of actions) {
-    for (const key in action.parameter_queries) {
-      action.parameter_queries[key] = await evaluateExpression(action.parameter_queries[key], {
-        objects: [payload],
-        pgConnection: PostgresDB,
+  for (const actionList of [actionsAsync, actionsSequential]) {
+    const status = Object.is(actionList, actionsAsync) ? 'Queued' : 'Processing'
+    // Add all actions to Action Queue
+    for (const action of actionList) {
+      // TODO - better error handling
+      await DBConnect.addActionQueue({
+        trigger_event: trigger_id,
+        template_id: templateId,
+        sequence: action.sequence,
+        action_code: action.code,
+        trigger_payload: payload,
+        parameter_queries: action.parameter_queries,
+        parameters_evaluated: {},
+        status,
       })
     }
-    // TODO - Error handling
-    // Write each Action with parameters to Action_Queue
-    await PostgresDB.addActionQueue({
-      trigger_event: payload.id,
-      action_code: action.code,
-      parameters: action.parameter_queries,
-      status: 'Queued',
-    })
+  }
+  await DBConnect.updateTriggerQueueStatus({ status: 'Actions Dispatched', id: trigger_id })
+
+  // Get sequential Actions from database
+  const actionsToExecute = await DBConnect.getActionsProcessing(templateId)
+
+  // Execute Actions one by one
+  for (const action of actionsToExecute) {
+    const actionPayload = {
+      id: action.id,
+      code: action.action_code,
+      trigger_payload: action.trigger_payload,
+      parameter_queries: action.parameter_queries,
+    }
+    await executeAction(actionPayload, actionLibrary)
```

Comment:
does this really happen synchronously ? I thought for loops like this would need to be for await

Reply:
Yeah, that's why I'm using a `for...of` loop rather than `forEach`. ForEach doesn't work well within an async function, but `for...of`/`for...in` does (as well as traditional `for (let i; i<n; i++)` loops.

Reply:
great